pkg/core/blockchain.go

```diff
@@ -393,6 +396,26 @@ func (bc *Blockchain) init() error {
 	return bc.updateExtensibleWhitelist(bHeight)
 }
 
+// InitOnRestore initializes blockchain before the state sync process for `p` height
+// can be started.
+func (bc *Blockchain) InitOnRestore(p uint32) error {
+	if !bc.config.P2PStateExchangeExtensions || p < uint32(bc.config.StateSyncInterval) {
+		return errors.New("invalid state sync point")
+	}
+	var height uint32
+	if p > bc.config.MaxTraceableBlocks+1 {
+		height = p - bc.config.MaxTraceableBlocks - 1
+	}
+	bc.lock.Lock()
+	defer bc.lock.Unlock()
+	bc.p = p
+	if height > bc.blockHeight {
+		// Need to provide proper (bc).BlockHeight() to external users while syncing.
+		bc.blockHeight = height
```

Comment:
This was an attempt to adapt the existing block addition mechanism to the state exchange process, but I got the idea, so it will be fixed by `Insole` maintenance.